File: javascript/triangle/triangle.js
export class Triangle {
  constructor(...sides) {
    this.sides = sides;
  }

  kind() {
    this.validateTriangle();

    if (this.isEquilateral()) { return 'equilateral'; }
    if (this.isIsosceles()) { return 'isosceles'; }
    if (this.isScalene()) { return 'scalene'; }
    throw new Error("Could not determine triangle type :(");
  }

  isEquilateral() {
    let [a,b,c] = this.sides;
    return a === b && a === c;
  }

  isIsosceles() {
    const [a,b] = this.sortedSides();
    return a === b;
  }

  isScalene() {
    const [a,b,c] = this.sides;
    return a !== b && a !== c && b !== c;
  }

  sortedSides() {
    return this.sides.slice(0).sort((a,b) => b - a);
  }

  validateTriangle() {
    this.validateNoZeroSides();
    this.validateTriangleInequality();
  }

  validateNoZeroSides() {
    if (Math.min(...this.sides) <= 0) {
      throw new Error("Sides cannot be zero");
    }
  }

  validateTriangleInequality() {
    const [longest, ...rest] = this.sortedSides();
    const sum = rest.reduce((acc, n) => n + acc);
    if (sum <= longest) {
      throw new Error("This is a degenerate triangle. Illegal!");
    }
  }
}

File: ruby/triangle/triangle.rb
class Triangle

  attr_accessor :sides

  def initialize(sides)
    @sides = sides
  end

  def equilateral?
    return false if illegal?
    sides.uniq.count == 1
  end

  def isosceles?
    return false if illegal?
    sides.uniq.count <= 2
  end

  def scalene?
    return false if illegal?
    sides.uniq.count == 3
  end

  private

  def illegal?
    return true if sides.any? {|side| side == 0}

    sides.each_with_index {|side, i|
      other_sides = sides.dup
      other_sides.delete_at(i)
      return true if side > other_sides.sum
    }

    false
  end

end
